Add Artifacts to Pkg (#1277)
JuliaLang/Pkg.jl: docs/src/artifacts.md

# [**8.** Artifacts](@id Artifacts)

> `Pkg` can reason about containers of binary data that are not Julia packages.  These chunks of data can contain platform-specific binaries, datasets, text, or any other kind of data that would be convenient to place within an immutable, lifecycled datastore.

Reviewer: ```suggestion
`Pkg` can install and manage containers of data that are not Julia packages.  These containers can contain platform-specific binaries, datasets, text, or any other kind of data that would be convenient to place within an immutable, life-cycled datastore.
```

These containers, (called "Artifacts") can be created locally, hosted anywhere, and automatically downloaded and unpacked upon installation of your Julia package.

> They are used to provide the binary dependencies for packages built with [`BinaryBuilder.jl`](https://github.com/JuliaPackaging/BinaryBuilder.jl).

Reviewer: ```suggestion
This mechanism is also used to provide the binary dependencies for packages built with [`BinaryBuilder.jl`](https://github.com/JuliaPackaging/BinaryBuilder.jl).
```

## `Artifacts.toml` files

> `Pkg` provides an API for working with artifacts, as well as a TOML file format for recording artifact usage in your packages, and to make automatic downloading of artifacts at package install time.

Reviewer: ```suggestion
`Pkg` provides an API for working with artifacts, as well as a TOML file format for recording artifact usage in your packages, and to automate downloading of artifacts at package install time.
```

> Artifacts are referred to by content-hash, or optionally by a name that is bound to a hash through an `Artifacts.toml` file.

Reviewer: ```suggestion
Artifacts can always be referred to by content hash, but are typically accessed by a name that is bound to a content hash in an `Artifacts.toml` file that lives in a project's source tree.
```

Note that each artifact contains both a `git-tree-sha1` and a `sha256` for each download entry.  This is to ensure that the downloaded tarball is secure before attempting to unpack it, as well as enforcing that all tarballs must expand to the same overall tree hash.

The `processed_output` artifact contains no `download` stanza, and so cannot be installed.
An artifact such as this would be the result of code that was previously run, generating a new artifact and binding the resultant hash to a name within this project.

## Using Artifacts

Artifacts can be manipulated using convenient APIs exposed from the `Pkg.Artifacts` namespace.
As a motivating example, let us imagine that we are writing a package that needs to load the [Iris machine learning dataset](https://archive.ics.uci.edu/ml/datasets/iris).
While we could just download the dataset during a build step into the package directory, and many packages currently do precisely this, that has some significant drawbacks:

> * First, it modifies the package directory, making package installation stateful, which we'd like to avoid.  In the future, we would like to reach the point where packages can be installed completely read-only, instead of being able to modify themselves after installation.

Reviewer: ```suggestion
* First, it modifies the package directory, making package installation stateful, which we want to avoid.
  In the future, we would like to reach the point where packages can be installed completely read-only, instead of being able to modify themselves after installation.

```

> * Second, the downloaded data is not shared across different versions of our package. If we have three different versions of the package installed for use by various projects, then we need three different copies of the data, even if it hasn't changed between those versions. Moreover, each time we upgrade or downgrade the package, unless we do something clever (and probably brittle), we will have to download the data again.

Reviewer: ```suggestion
* Second, the downloaded data is not shared across different versions of our package.
  If we have three different versions of the package installed for use by various projects, then we need three different copies of the data, even if it hasn't changed between those versions.
  Moreover, each time we upgrade or downgrade the package, unless we do something clever (and probably brittle), we have to download the data again.
```

* **Hash-aware** functions deal with content-hashes and essentially nothing else. These methods allow you to query whether an artifact exists, what its path is, to verify that an artifact satisfies its content hash on-disk, etc.  Hash-aware functions include: `artifact_exists()`, `artifact_path()`, `remove_artifact()`, `verify_artifact()` and `archive_artifact()`.  Note that in general you should not use `remove_artifact()` and should instead use `Pkg.gc()` to cleanup artifact installations.

* **Name-aware** functions deal with bound names within an `Artifacts.toml` file, and as such, typically require both a path to an `Artifacts.toml` file as well as the artifact name.  Name-aware functions include: `artifact_meta()`, `artifact_hash()`, `bind_artifact!()`, `unbind_artifact!()`, `download_artifact()` and `ensure_artifact_installed()`.

* **Utility** functions deal with miscellaneous aspects of artifact life, such as `create_artifact()`, `ensure_all_artifacts_installed()`, and even the `@artifact_str` string macro.

For a full listing of docstrings and methods, see the [Artifacts Reference](@ref) section.

## Overriding artifact locations

> It is occasionally convenient to be able to override the location of an artifact.

Reviewer: ```suggestion
It is occasionally necessary to be able to override the location and content of an artifact.
```

> A common usecase is a restrictive computing environment where certain versions of a binary dependency must be used.

Reviewer: ```suggestion
A common use case is a computing environment where certain versions of a binary dependency must be used, regardless of what version of this dependency a package was published with.
```

While a typical Julia configuration would download, unpack and link against a generic library, a system administrator may wish to disable this and instead use a library already installed on the local machine.
To enable this, `Pkg` supports a per-depot `Overrides.toml` file placed within the `artifacts` depot directory (e.g. `~/.julia/artifacts/Overrides.toml` for the default user depot) that can override the location of an artifact either by content-hash or by package UUID and bound artifact name.

> Additionally, the destination location can be either an absolute path, or a new artifact content-hash, to support sysadmins in creating new, custom artifacts, then overriding other packages to use that new artifact.

Reviewer: ```suggestion
Additionally, the destination location can be either an absolute path, or a replacement artifact content hash.
This allows sysadmins to creating their own artifacts which they can then use by overriding other packages to use the new artifact.
```

# [**8.** Artifacts](@id Artifacts)

`Pkg` can install and manage containers of data that are not Julia packages.  These containers can contain platform-specific binaries, datasets, text, or any other kind of data that would be convenient to place within an immutable, life-cycled datastore.
These containers, (called "Artifacts") can be created locally, hosted anywhere, and automatically downloaded and unpacked upon installation of your Julia package.
This mechanism is also used to provide the binary dependencies for packages built with [`BinaryBuilder.jl`](https://github.com/JuliaPackaging/BinaryBuilder.jl).

## `Artifacts.toml` files

`Pkg` provides an API for working with artifacts, as well as a TOML file format for recording artifact usage in your packages, and to automate downloading of artifacts at package install time.
Artifacts can always be referred to by content hash, but are typically accessed by a name that is bound to a content hash in an `Artifacts.toml` file that lives in a project's source tree.
An example `Artifacts.toml` file is shown here as an example:

> ```TOML

Reviewer 2: We use
```suggestion
```toml
```
in other places but not sure whats correct actually.

Author: I think they both work

> # Example Artifacts.toml file

Reviewer 2: ```suggestion
```

Author: I would rather have that there, so that users that are scanning through quickly can see that (a) comments in TOML file works, and (b) they know the filename of this file.
